server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -618,9 +621,20 @@ protected void onShardResult(Result result, SearchShardIterator shardIt) {
         if (logger.isTraceEnabled()) {
             logger.trace("got first-phase result from {}", result != null ? result.getSearchShardTarget() : null);
         }
+        this.setPhaseResourceUsages();
         results.consumeResult(result, () -> onShardResultConsumed(result, shardIt));
     }
 
+    public void setPhaseResourceUsages() {
+        ThreadContext threadContext = searchRequestContext.getThreadContextSupplier().get();
+        if (threadContext != null) {
+            List<String> taskResourceUsages = threadContext.getResponseHeaders().get(TASK_RESOURCE_USAGE);
```

Comment:
We are exposing arbitrary headers with the whole thread context. Probably better to supply only the `TASK_RESOURCE_USAGE` headers, since that is all we need

Reply:
That's a good point! let me refactor this to make it a supplier of the `TASK_RESOURCE_USAGE` String only.